### PR TITLE
[SYCL-MLIR] Version loop to check for different accessors

### PR DESCRIPTION
**Pull request by whitneywhtsang**
For example, given the loop below:
```
for(size_t i = 0; i < 8; i++) {
  A[i] += B[0];
}
```
`B[0]` can be hoisted out of the loop, if `A` and `B` are not aliased. 
Conservatively, we can version the loop with `(&A[A.get_range()-1]+1) <= &B[sysl::id{}] || &A[sycl::id{}] >= (&B[B.get_range()-1]+1)` to ensure the two accessors don't overlap. 

Currently, loop versioning it is disabled by default, and can be enabled by `-enable-sycl-accessor-versioning`.

### CODE REVIEW

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:823]: Will handle 32 bit pointer type on a different PR.

File: polygeist/include/mlir/Dialect/Polygeist/Transforms/Passes.td
Comment (victor-eds) [inline polygeist/include/mlir/Dialect/Polygeist/Transforms/Passes.td:112]: Should we add the `sycl` dialect also here as we will create operations from it?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:729]: Maybe assert?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:762]: Maybe assert?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:870]: NIT: Instead of relying on the body of the type, we could be creating a new `IDType` here... It's also true we'd need to be creating the body in an ad-hoc way, so I don't have a strong preference.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:906]: Wouldn't the end be `accSub`?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:914]: NIT: Have `AccessorPair` as two different values and check nullity in `empty`

**Review by victor-eds — COMMENTED** [review]
This is a good first step. What changes would we need to make this support N accessors?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:900]: I have to subtract 1 here, and add 1 to the pointer after, or else we are accessing out of bound.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:906]: not really, let say accessor A has get_range = {3, 3}, the end pointer should be &A[{2, 2}] + 1, and not &A[{3, 3}], as that would be out of bound, and &A[{2, 2}] would be the beginning of the last element.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:914]: I picked to use a pair, because If we extent it to allow more than one pair, then we need something like a vector of `AccessorPair`. WDYT?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:762]: We should just not version if there is not candidates. IMO, shouldn't assert.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:729]: This is a temporary limitation, which only allow one candidate. Either we return false here in `isVersionCandidate`, or we need the user of `isVersionCandidate` to check before calling `isVersionCandidate`, which means exposing `empty()` too. I personally prefer the first way, do you prefer the second way?

**Comment by whitneywhtsang** [comment]
> This is a good first step. What changes would we need to make this support N accessors?

It shouldn't be too bad to support N accessors without a cost model, we just need to remove the `if (!empty()) return false;` in `isVersionCandidate()`, change `AccessorPair` to a vector of `AccessorPair`s, and concatenate all conditions with `and` when generating the version condition. The harder part is to have a cost model to decide which accessor pairs to version.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (victor-eds) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:906]: True. Can we not use `polygeist::SubIndexOp`, but building the input ID taking that into account?

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (victor-eds) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:914]: True!

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (victor-eds) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:729]: No, this is good to me :+1:

File: polygeist/include/mlir/Dialect/Polygeist/Transforms/Passes.td
Comment (whitneywhtsang) [reply polygeist/include/mlir/Dialect/Polygeist/Transforms/Passes.td]: done.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:900]: let's continue the discussion in https://github.com/intel/llvm/pull/8935#discussion_r1157136523.

**Comment by etiotto** [comment]
> > This is a good first step. What changes would we need to make this support N accessors?
> 
> It shouldn't be too bad to support N accessors without a cost model, we just need to remove the `if (!empty()) return false;` in `isVersionCandidate()`, change `AccessorPair` to a vector of `AccessorPair`s, and concatenate all conditions with `and` when generating the version condition. The harder part is to have a cost model to decide which accessor pairs to version.

I propose this PR limits itself to handle versioning one descriptor. The key is the cost model because hoisting a value that require the loop to be versioned can lead to other opportunities. So the cost model should attempt to maximize the number of hoistable value while at the same time minimize the number of loop versioned. That is tricky so I think is best left to a future PR.

**Review by victor-eds — APPROVED** [review]

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (victor-eds) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:902]: Subtract 1 from the last dimension of `get_range()` when building the offset ID

Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:902]: oic! do you mean plus 1 from the last dimension?

Comment (victor-eds) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:902]: Yes, that's what I meant!

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:942]: This is strange because looks like the `SYCLAccessorVersionBuilder` represents both the class that create the versioned loop as well as the candidate operations that require versioning in order to be hoisted. Can we have 2 classes, one to represent the builder and one to represent the candidate operation that require versioning ? 

I think we need: 
 - a loop versioning builder which is responsible to version a loop, it should receive the Loop to be versioned and the (aggregated) condition to use to version the loop.
 - a LICM candidate, which should know the condition it needs in order for itself to be hoisted. 

Once we have those 2 classes we need 'something' that collect all candidates and create the aggregated condition to give to the loop version builder.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:780]: auto -> Value

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:905]: emptu() -> empty() const

**Review by etiotto — COMMENTED** [review]
Code to generate the check LGTM (after addressing the comments ion the PR).
I left comments on how to structure the various classes required to version the loop, represent a candidate operation that require versioning, etc...

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:792]: Added `createMethodOp`.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:869]: Keeping to ensure we are creating the same IDType as the one used in AccessorType.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:934]: Created https://github.com/intel/llvm/pull/8950 to add a `LICMCandidate` class first.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:1172]: The current function `moveLoopInvariantCode` should (after collecting candidates) just concern itself with the mechanics of hoisting the candidates (that might require versioning the loop).  This code should be placed in some "pruning/filtering" function that decide which candidates to remove based on a set of heuristics (e.g. max number of versioning condition equal to N).

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:751]: that require not to overlap -> that are required to not overlap

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:765]: SmallVector -> const SmallVectorImpl

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:1168]: SmallVector<AccessorPairType> --> const SmallVector<AccessorPairType>&

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:1170]: if (AccessorPairs.size() != 0)  --> if (!AccessorPairs.empty())

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:905]: I'd like the static function to be moved at the top of the file so that (a) it can be used throughout the file if we need to and (b) is easy to locate all the static function(s) declared in this file.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (etiotto) [inline polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:1122]: Suggest: Some candidate operations require the loop to be versioned to be hoisted, if the loop is not versioned because of heuristic considerations (e.g. exceeded the versioning limit) we have to filter out the operations that depend upon them.

**Review by etiotto — APPROVED** [review]
LGTM. I left some inline comment that should be addressed before merging.

File: polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/LICM.cpp:905]: It is currently with all other static functions, and it is the first static function, so I think the current position is fine.

**Pull request merged**
